src/rviz/validate_quaternions.h

```diff
@@ -40,104 +40,146 @@
 namespace rviz
 {
 
-inline float quaternionNorm2(float w, float x, float y, float z)
+// Tolerance for acceptable quaternion normalization (same as in tf)
+static double QUATERNION_NORMALIZATION_TOLERANCE = 10e-3;
+
+inline float quaternionNorm2( float w, float x, float y, float z )
 {
   return w * w + x * x + y * y + z * z;
 }
 
-inline bool validateQuaternion(float w, float x, float y, float z)
+inline bool validateQuaternions( float w, float x, float y, float z )
 {
-  return std::abs(quaternionNorm2(w, x, y, z) - 1.0f) < 10e-3f;
+  if ( 0.0f == x && 0.0f == y && 0.0f == z && 0.0f == w )
+  {
+    // Allow null quaternions to pass because they are common in uninitialized ROS messages.
+    return true;
+  }
+  float norm2 = quaternionNorm2( w, x, y, z ); 
+  bool is_normalized = std::abs( norm2 - 1.0f ) < QUATERNION_NORMALIZATION_TOLERANCE;
+  ROS_DEBUG_COND_NAMED( !is_normalized, "quaternions", "Quaternion [x: %.3f, y: %.3f, z: %.3f, w: %.3f] not normalized. "
+                        "Magnitude: %.3f", x, y, z, w, std::sqrt(norm2) );
+  return is_normalized;
 }
 
-inline bool normalizeQuaternion(float& w, float& x, float& y, float& z)
+inline float normalizeQuaternion( float& w, float& x, float& y, float& z )
 {
-  float norm2 = quaternionNorm2(w, x, y, z);
-  if (norm2 < 10e-3f)
-    return false;
-  norm2 = std::sqrt(norm2);
+  if ( 0.0f == x && 0.0f == y && 0.0f == z && 0.0f == w )
+  {
+    w = 1.0f;
+    x = y = z = 0.0f;
```

Comment:
This line is superfluous. x,y,z are zero already.

Reply:
gah, thanks for pointing that out: df6dfd4